# A quick recap on Pandas DataFrames

Pandas DataFrames are mutable two-dimensional structures of data with labeled axes where: 
* each row represents a different observation
* each column represents a different variable

As always, we first need to import the Pandas module:

# New Section

In [3]:
import pandas as pd

## 1.&nbsp; Import a csv file to DataFrame

Most of the time, you will not be creating DataFrames yourself, but importing (or "reading") data from a csv file or a database into a pandas DataFrame. It's easy to do with pandas' read functions. We will read one of the datasets from the Eniac project, which we have stored in a Google Drive folder:

In [9]:
url = "https://drive.google.com/file/d/1FYhN_2AzTBFuWcfHaRuKcuCE6CWXsWtG/view?usp=sharing" # orderlines.csv
path = "https://drive.google.com/uc?export=download&id="+url.split("/")[-2]
df = pd.read_csv(path)
df


id  id_order  product_id  product_quantity      sku unit_price  \
0       1119109    299539           0                 1  OTT0133      18.99   
1       1119110    299540           0                 1  LGE0043     399.00   
2       1119111    299541           0                 1  PAR0071     474.05   
3       1119112    299542           0                 1  WDT0315      68.39   
4       1119113    299543           0                 1  JBL0104      23.74   
...         ...       ...         ...               ...      ...        ...   
293978  1650199    527398           0                 1  JBL0122      42.99   
293979  1650200    527399           0                 1  PAC0653     141.58   
293980  1650201    527400           0                 2  APP0698       9.99   
293981  1650202    527388           0                 1  BEZ0204      19.99   
293982  1650203    527401           0                 1  APP0927      13.99   

                       date  
0       2017-01-01 00:07:19  
1       2017-01-01 00:19:45  
2       2017-01-01 00:20:57  
3       2017-01-01 00:51:40  
4       2017-01-01 01:06:38  
...                     ...  
293978  2018-03-14 13:57:25  
293979  2018-03-14 13:57:34  
293980  2018-03-14 13:57:41  
293981  2018-03-14 13:58:01  
293982  2018-03-14 13:58:36  

[293983 rows x 7 columns]

## 2.&nbsp; DataFrame dimensions

With the [DataFrame.shape](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.shape.html) attribute we can calculate the dimensions (number of rows and columns) of the DataFrame.

In [5]:
df.unit_price.count()

293983

As a result we obtain a [tuple](https://www.w3schools.com/python/python_tuples.asp) where the first element is the number of rows, which in our case is 293983, while the second element is the number of columns, which in our example was 7.

In [8]:
nrows = df.shape[0]
ncols = df.shape[1]
print("The number of rows is", nrows)
print("The number of columns is", ncols)

The number of rows is 293983
The number of columns is 7


In [11]:
df.plot.scatter(x = "sku", y= "unit_price");

Error in callback <function flush_figures at 0x7f3bffc14200> (for post_execute):


KeyboardInterrupt: ignored

[DataFrame.size](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.size.html) returns the total number of values that the DataFrame has (the number of rows multiplied by the number of columns):

In [7]:
df.size

2057881

We can check if the `.size` and `.shape` agree

In [6]:
df.shape[0] * df.shape[1] == df.size

True

With the [DataFrame.ndim](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.ndim.html) attribute we calculate the number of dimensions that the DataFrame has. This will always be 2, as it consists of rows and columns.

In [ ]:
df.ndim

## 3.&nbsp; DataFrames exploration

The [DataFrame.head()](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.head.html) and [DataFrame.tail()](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.tail.html) methods are used to display the first or last rows of the DataFrame. Looking at the raw data is a great way to get a grasp of what's in there. By default, 5 rows will be shown, but you can change that:

In [ ]:
df.head()

In [ ]:
df.head(9)

In [ ]:
df.tail()

The methods [DataFrame.info()](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.info.html), [DataFrame.describe()](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.describe.html) and [DataFrame.nunique()](https://pandas.pydata.org/docs/reference/api/pandas.core.groupby.DataFrameGroupBy.nunique.html) give a general overview of what's in the DataFrame:

`.info()` tells us about how pandas sees the data - how it is stored, whether there's any missing values, and how many columns and rows we have.

In [ ]:
df.info()

`.describe()` gives us an overview of the [descriptive statistics](https://www.scribbr.com/statistics/descriptive-statistics/) for the numerical columns of our DataFrame.

In [ ]:
df.describe()

In [ ]:
df.nunique()

If we wish, we can also view the unique values counted by `series.nunique()`. The [.unique()](https://pandas.pydata.org/docs/reference/api/pandas.Series.unique.html) method returns the unique values from a column as a numpy array, which can be indexed with `[]`:

In [ ]:
df["sku"].unique()[:10]

The [DataFrame.isna()](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.isna.html) method returns a boolean for each value: `True` if that value is "missing" (which is represented as `NaN` in numpy and pandas) and `False` if the value is not missing:

In [ ]:
df.isna()

We can then use [DataFrame.sum()](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.sum.html) to add up all these booleans for each column, and count how many missing values are there in the DataFrame, since `True` is interpreted as `1` and `False` as `0`:

In [ ]:
df.isna().sum()

[DataFrame.duplicated()](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.duplicated.html) also returns a boolean output, but in this case just one value per row: `True` if that row is duplicated and `False` if it's not. Again, using `sum()` allows us to count how many `True` values (i.e. duplicated rows) are there in total:

In [ ]:
df.duplicated().sum()

[DataFrame.nlargest(n, columns)](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.nlargest.html) will return the top `n` rows with the largest value for whatever column we specify in `columns`. Below, we see the rows with the largest product quantity values:

In [ ]:
df.nlargest(5, "product_quantity")

[DataFrame.nsmallest()](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.nsmallest.html) does the same, for the smallest values:

In [ ]:
df.nsmallest(5, "product_quantity")

So far, we have seen two ways to explore DataFrames:

* **Attributes:** `.shape`, `.size` and `.ndim`, and others. They are written without parentheses and give you raw "metadata" about the DataFrame you are calling them on.
* **Methods:** `.head()`, `.describe()` and `.isna()`, and others. They are written with parentheses and perform some sort of calculation, transformation or aggregation. A method is like a function that is tied to a specific object type.

DataFrames have a lot of attributes and methods and may not be obvious whether something belongs to one type or the other. Whenever in doubt, check [the documentation](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.html?highlight=DataFrame#pandas.DataFrame). If you scroll down past the list of examples you will find two sections listing all the attributes and methods.

## 4.&nbsp; Select Columns

Given a DataFrame, we can select a particular column in several ways:

* Indicating the name of the column between square brackets, `[]`
* With the `.loc[]` attribute (by name or tag)
* With the `.iloc[]` attribute (by position)

Plain square brackets `[]` are used to just view a column if you know its name and **don't want to modify it**:

In [ ]:
df["id_order"]

`.loc[]` takes two arguments: `[rows, columns]`. Passing `:` to the rows argument means "grabbing all the rows", which allows you to select a whole column, if you know its name. This method is more flexible as you will see in the future, and allows you to modify the data. 

We recommend using `loc[]` as the primary option for selecting data.

In [ ]:
df.loc[:, "id_order"]

`.iloc[]` works similarly to `.loc[]`, but it only accepts integers, which represent the positions of the rows and columns:

In [ ]:
df.iloc[:, 1]

### 4.1.&nbsp; Select multiple columns

If we want to select more than one column, we can do it with all the options listed above, with slight modifications in some cases:

> **Note:** we pass a list inside of the `[]`, a list is also represented by `[]`, hence why we have 2 sets of square brackets

In [ ]:
df[["id_order","sku"]]

In [ ]:
df.loc[:, ["id_order", "sku"]]

With `.loc[]` and `:` you can select all columns between two columns you specify.

In [ ]:
df.loc[:, "id_order":"sku"]

In [ ]:
df.iloc[:, [1, 4]]

In [ ]:
df.iloc[:, 1:5]

## 5.&nbsp; Select Rows

Selecting rows is easy if you know how to select columns. You have two options:

* With `.loc[]` (by name or tag)
* With `.iloc[]` (by position)

Selecting a single row returns a pandas Series (a 1-dimensional object):

In [ ]:
df.loc[0, :]

With `.loc[]`, rows are selected by their index name:

In [ ]:
df.loc[0:3, :]

If we change the index and set it to the `id` column, now the first rows can not be selected the same way:
> **Note:** `inplace = True` is the same as doing `df = df.set_index("id")`, i.e. it modifies the DataFrame

In [ ]:
df.set_index("id", inplace=True)

Look, the same code as above no longer works. The index has been changed to the `id` column, if we want to use `.loc` we now need to use the names/numbers from `id`

In [ ]:
df.loc[0:3, :]

In [ ]:
df.loc[1119109:1119112, :]

You can always tell the index column as it will be in bold

With the `.iloc[]` method we don't need to know the row names to select rows at a certain position. Every row and column is instead nominally indexed, starting from the number 0:

In [ ]:
df.iloc[0:4]

We can select the last observation with the method `.iloc[]` by using negative numbers

In [ ]:
df.iloc[-1]

We can, whenever necessary, reset the index to a set of numbers starting from 0. This would also make the `id` column a normal column once again

In [ ]:
df.reset_index(inplace=True)

Indexing can get tricky sometimes, it's ok to take some time to get used to the methods we presented, and it's ok to have some trouble selecting the rows and columns you need. For an exhaustive guide on Pandas indexing, check out this link: https://pandas.pydata.org/docs/user_guide/indexing.html#indexing 

## 6.&nbsp; Drop Columns

The `.drop()` method allows us to delete the rows or columns.

> **Note:** Again, if we want to directly apply the changes to the original DataFrame, we need to indicate `inplace = True`. Otherwise, we are getting as an output just a "view" of how the DataFrame looks like after the drop, but the original DataFrame remains unchanged.

> **Note:** `axis=1` means we want to drop a column, not a row. For rows we would use `axis=0`

In [ ]:
df.drop(["unit_price"], axis=1)

See how our output above shows no column `unit_price`. Then look at the output below, `unit_price` is magically back. We didn't use `inplace=True` above, which means that our output displayed what we asked, but didn't alter the original DataFrame. Be careful of this, Pandas only changes data if you explicitly tell it to do so.

## 7.&nbsp; Filter rows based on conditions

Using `.loc` to filter for rows that have a `product_quantity greater than 100`

In [ ]:
df.loc[df["product_quantity"] > 100, :]

The `.query()` method can be useful for this purpose. Some of you may prefer it as it resembles SQL syntax. You can use any **Python Comparison Operators** you want inside the query method (find more information on this [link](https://www.w3schools.com/python/python_operators.asp)).
> **Note:** `.query()` only works when what's contained in the `" "` does not contain any white space. This can affect both column names and cell values. For example if you have a column `a b` and you wish to query it `df.query("a b == 5")`, this would throw up an error. For column names you should always be using an underscore `_` not a space

# New Section

In [ ]:
df.query("product_quantity > 100")

The `isin()` method is very useful to find rows that match any of the values you have **in a list**. For example, here we are searching for rows where its `sku` matches any of the 2 sku's we listed:

In [ ]:
# find out a column that contains a list
df["sku"].isin(["JBL0104", "ADN0039"])

This expression can be used inside of `[]` or `.loc[]` to filter the rows that have a `True` value. This is called "boolean indexing" and it is really useful:

In [ ]:
df.loc[df["sku"].isin(["JBL0104", "ADN0039"]), :]

Pandas compresses large outputs in Colab/Jupyter Notebooks. If you want to see more rows, you can change the options:
> **Note:** after running this cell, run the code above again and you will see all the rows

In [ ]:
pd.options.display.max_rows = 100

For a complete list of all settings and options that can be tweaked, check out this: https://pandas.pydata.org/pandas-docs/stable/user_guide/options.html

#### 7.1.&nbsp; Modifying a DataFrame & the `.copy()` method

Let's take a small sample of data from our DataFrame:

In [ ]:
sample = df.iloc[:3,:]
sample

Now we pick a single cell from the sample we took and we assign a new value to it. A warning already tells us that this is a risky thing to do:

In [ ]:
sample.iloc[0,4] = "NEW VALUE HERE"

We can see the new value on the `sample` we took:

In [ ]:
sample

..and, maybe to your surprise, we can see that the new value is also present on the original `df`!

In [ ]:
df.head()

When you take a sample of data using `.loc[]` or `iloc[]` and assign it to a new object, the new object is just a "tag" pointing to the very same data as the original DataFrame points to. We can avoid this using the method `.copy()`

In [ ]:
import pandas as pd 
url = "https://drive.google.com/file/d/14Y7g5ITyf6LMyPoKc9wr010V9StaCUux/view?usp=sharing" # orderlines.csv
path = "https://drive.google.com/uc?export=download&id="+url.split("/")[-2]
df = pd.read_csv(path)

In [ ]:
sample = df.iloc[:3,:].copy()
sample.iloc[0,4] = "NEW VALUE HERE"
sample

In [ ]:
df.head(3)

As you can see, now it has not been modified.

# CHALLENGES

## Challenge 1: 
How many different unit prices does the product with the sku JBL0104 have? Combine a pandas filter method with the method `.nunique()`.

In [ ]:
# your code here

## Challenge 2: 
List the (unique) items that were sold in the order with the id_order 385921.

In [ ]:
# your code here

## Challenge 3: 
Consider the products with the skus APP2431 and APP2348. Find out in how many orders they were present.

In [ ]:
# your code here

## Challenge 4: 
Create a new DataFrame, `df_50`, with all the rows that have a product quantity higher than 500 and only the columns `id`, `product_id`, `product_quantity` and `sku`. Be sure to use the method `.copy()`. Once the new DataFrame is created, modify the column "product_quantity" to "quantity", and "sku" to "product_code". To do so, you can use the method `.rename()` or assign a list of new names to the attribute `.columns`.

Creating `df_50`

In [ ]:
# your code here

Renaming columns of `df_50`

In [ ]:
# your code here